xarray/backends/zarr.py

```diff
@@ -189,20 +189,38 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name, safe_chunks):
     # TODO: incorporate synchronizer to allow writes from multiple dask
     # threads
     if var_chunks and enc_chunks_tuple:
-        for zchunk, dchunks in zip(enc_chunks_tuple, var_chunks, strict=True):
-            for dchunk in dchunks[:-1]:
+        base_error = (
+            f"Specified zarr chunks encoding['chunks']={enc_chunks_tuple!r} for "
+            f"variable named {name!r} would overlap multiple dask chunks {var_chunks!r}. "
+            f"Writing this array in parallel with dask could lead to corrupted data."
+            f"Consider either rechunking using `chunk()`, deleting "
+            f"or modifying `encoding['chunks']`, or specify `safe_chunks=False`."
+        )
+
+        for zchunk, dchunks, interval in zip(
+            enc_chunks_tuple, var_chunks, region, strict=True
+        ):
+            if not safe_chunks or len(dchunks) <= 1:
```

Comment:
Re "so would this PR be ready then" — am I right in thinking the proposed code changes the behavior such that a single partial chunk can be written on `mode="r+"`? I don't think we want that yet (though am not confident, it seems unlikely to be a big problem, but still possibly confusing)

Reply:
Yes, the code is going to allow writing on a single partial chunk on mode="r+".
From my perspective and based on the docs of the to_zarr method, the new behavior would not be confusing, "“r+” means modify existing array values only (raise an error if any metadata or shapes would change)", and writing in partial chunk means modify the existing array values.

Reply:
But it breaks the rule we discussed above, re it being safe writing to different indexes from distributed processes.

It's _unlikely_, because it's rare to write to a single partial chunk from distributed processes. But is it worth the complication, relative to forcing people to pass `mode="a"`?

Reply:
@max-sixty I sent the following PR https://github.com/pydata/xarray/pull/9527 which contains all the changes requested, if you want I can add everything here or close this one, whatever you prefer.